crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,343 @@ impl MeshletMesh {
+                );
+            }
+
+            simplification_queue = next_lod_start..meshlets.len();
+            lod_level += 1;
+        }
 
-        // Assemble into the final asset
-        let mut total_meshlet_triangles = 0;
-        let meshlets = meshopt_meshlets
+        // Convert meshopt_Meshlet data to a custom format
+        let bevy_meshlets = meshlets
             .meshlets
             .into_iter()
-            .map(|m| {
-                total_meshlet_triangles += m.triangle_count as u64;
-                Meshlet {
-                    start_vertex_id: m.vertex_offset,
-                    start_index_id: m.triangle_offset,
-                    triangle_count: m.triangle_count,
-                }
+            .map(|m| Meshlet {
+                start_vertex_id: m.vertex_offset,
+                start_index_id: m.triangle_offset,
+                triangle_count: m.triangle_count,
             })
             .collect();
 
         Ok(Self {
-            total_meshlet_triangles,
+            worst_case_meshlet_triangles,
             vertex_data: vertex_buffer.into(),
-            vertex_ids: meshopt_meshlets.vertices.into(),
-            indices: meshopt_meshlets.triangles.into(),
-            meshlets,
-            meshlet_bounding_spheres,
+            vertex_ids: meshlets.vertices.into(),
+            indices: meshlets.triangles.into(),
+            meshlets: bevy_meshlets,
+            bounding_spheres: bounding_spheres.into(),
         })
     }
 }
 
+fn validate_input_mesh(mesh: &Mesh) -> Result<Cow<'_, [u32]>, MeshToMeshletMeshConversionError> {
+    if mesh.primitive_topology() != PrimitiveTopology::TriangleList {
+        return Err(MeshToMeshletMeshConversionError::WrongMeshPrimitiveTopology);
+    }
+
+    if mesh.attributes().map(|(id, _)| id).ne([
+        Mesh::ATTRIBUTE_POSITION.id,
+        Mesh::ATTRIBUTE_NORMAL.id,
+        Mesh::ATTRIBUTE_UV_0.id,
+        Mesh::ATTRIBUTE_TANGENT.id,
+    ]) {
+        return Err(MeshToMeshletMeshConversionError::WrongMeshVertexAttributes);
+    }
+
+    match mesh.indices() {
+        Some(Indices::U32(indices)) => Ok(Cow::Borrowed(indices.as_slice())),
+        Some(Indices::U16(indices)) => Ok(indices.iter().map(|i| *i as u32).collect()),
+        _ => Err(MeshToMeshletMeshConversionError::MeshMissingIndices),
+    }
+}
+
+fn compute_meshlets(indices: &[u32], vertices: &VertexDataAdapter) -> Meshlets {
+    let mut meshlets = build_meshlets(indices, vertices, 64, 64, 0.0);
+
+    for i in 0..meshlets.len() {
+        let meshlet = meshlets.meshlets[i];
+        #[allow(unsafe_code)]
+        #[allow(clippy::undocumented_unsafe_blocks)]
+        unsafe {
+            meshopt_optimizeMeshlet(
+                &mut meshlets.vertices[meshlet.vertex_offset as usize],
+                &mut meshlets.triangles[meshlet.triangle_offset as usize],
+                meshlet.triangle_count as usize,
+                meshlet.vertex_count as usize,
+            );
+        }
```

Comment:
Can we add a safe wrapper for this to the `meshopt` crate instead? It looks like the safe wrapper would be trivial, and that would let us avoid messing with the unsafe code settings in this crate.

Reply:
Yeah it's not published yet though, so for now, unsafe it is.